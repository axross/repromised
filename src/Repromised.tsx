import { PureComponent, ReactNode } from 'react';

type Children<Value> = (value: Value, isProcessing: boolean) => ReactNode;

type Props<Value> = {
  promise: () => Promise<Value>;
  initial: Value;
  then?: (value: Value) => void;
  catch?: (err: Error) => void;
  beforeResolve?: () => void;
  children?: Children<Value>;
};

type State<Value> = {
  value: Value;
  isProcessing: boolean;
};

class Repromised<Value> extends PureComponent<Props<Value>, State<Value>> {
  public componentDidMount() {
    this.setState({ isProcessing: true });

    if (this.props.beforeResolve) {
      this.props.beforeResolve();
    }

    this.props
      .promise()
      .then(value => {
        this.setState({ value, isProcessing: false });

        if (this.props.then) {
          this.props.then(value);
        }
      })
      .catch(err => {
        if (this.props.catch) {
          this.props.catch(err);
        }

        this.setState({ isProcessing: false });
      });
  }

  public render() {
    if (this.props.children) {
      return this.props.children(this.state.value, this.state.isProcessing);
    }

    return null;
  }

  public constructor(props: Props<Value>) {
    super(props);

    this.state = {
      value: props.initial,
      isProcessing: false,
    };
  }
}

export default Repromised;
