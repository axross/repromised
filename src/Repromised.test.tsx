import { createElement } from "react";
import * as TestRenderer from "react-test-renderer";
import Repromised, { Status } from "./Repromised";

describe("<Repromised>", () => {
  describe("when props.promise resolves with a value", () => {
    test("calls props.children with the snapshot which expresses the promise is pending at 1st call", done => {
      const children = jest.fn(_ => null);
      const promiseFunction = jest.fn(() => Promise.resolve("PROMISED"));

      TestRenderer.create(
        <Repromised promise={promiseFunction} initial={"INITIAL"}>
          {children}
        </Repromised>
      );

      setTimeout(() => {
        expect(children.mock.calls[0][0].status).toBe(Status.Pending);
        expect(children.mock.calls[0][0].value).toBe("INITIAL");
        expect(children.mock.calls[0][0].error).toBe(null);
        expect(children.mock.calls[0][0].isLoading).toBe(true);
        expect(children.mock.calls[0][0].requireValue).toBe("INITIAL");

        done();
      }, 5);
    });

    test("calls props.children with the snapshot which expresses the promise is fulfilled at 2st call", done => {
      const children = jest.fn(_ => null);
      const promiseFunction = jest.fn(() => Promise.resolve("PROMISED"));

      TestRenderer.create(
        <Repromised promise={promiseFunction} initial={"INITIAL"}>
          {children}
        </Repromised>
      );

      setTimeout(() => {
        expect(children.mock.calls[1][0].status).toBe(Status.Fulfilled);
        expect(children.mock.calls[1][0].value).toBe("PROMISED");
        expect(children.mock.calls[1][0].error).toBe(null);
        expect(children.mock.calls[1][0].isLoading).toBe(false);
        expect(children.mock.calls[1][0].requireValue).toBe("PROMISED");

        done();
      }, 5);
    });
  });

  describe("when props.promise rejects with an error", () => {
    test("calls props.children with the snapshot which expresses the promise is pending at 1st call", done => {
      const children = jest.fn(_ => null);
      const error = new Error("REJECTED");
      const promiseFunction = jest.fn(() => Promise.reject<any>(error));

      TestRenderer.create(
        <Repromised promise={promiseFunction} initial={"INITIAL"}>
          {children}
        </Repromised>
      );

      setTimeout(() => {
        expect(children.mock.calls[0][0].status).toBe(Status.Pending);
        expect(children.mock.calls[0][0].value).toBe("INITIAL");
        expect(children.mock.calls[0][0].error).toBe(null);
        expect(children.mock.calls[0][0].isLoading).toBe(true);
        expect(children.mock.calls[0][0].requireValue).toBe("INITIAL");

        done();
      }, 5);
    });

    test("calls props.children with the snapshot which expresses the promise is rejected at 2st call", done => {
      const children = jest.fn(_ => null);
      const error = new Error("REJECTED");
      const promiseFunction = jest.fn(() => Promise.reject<any>(error));

      TestRenderer.create(
        <Repromised promise={promiseFunction} initial={"INITIAL"}>
          {children}
        </Repromised>
      );

      setTimeout(() => {
        expect(children.mock.calls[1][0].status).toBe(Status.Rejected);
        expect(children.mock.calls[1][0].value).toBe(null);
        expect(children.mock.calls[1][0].error).toBe(error);
        expect(children.mock.calls[1][0].isLoading).toBe(false);
        expect(() => children.mock.calls[1][0].requireValue).toThrow();

        done();
      }, 5);
    });
  });

  describe("props.initial is omitted", () => {
    describe("when props.promise resolves with a value", () => {
      test("calls props.children with the snapshot which expresses the promise is pending at 1st call", done => {
        const children = jest.fn(_ => null);
        const promiseFunction = jest.fn(() => Promise.resolve("PROMISED"));

        TestRenderer.create(
          <Repromised promise={promiseFunction}>{children}</Repromised>
        );

        setTimeout(() => {
          expect(children.mock.calls[0][0].status).toBe(Status.Pending);
          expect(children.mock.calls[0][0].value).toBe(null);
          expect(children.mock.calls[0][0].error).toBe(null);
          expect(children.mock.calls[0][0].isLoading).toBe(true);
          expect(() => children.mock.calls[0][0].requireValue).toThrow();

          done();
        }, 5);
      });

      test("calls props.children with the snapshot which expresses the promise is fulfilled at 2st call", done => {
        const children = jest.fn(_ => null);
        const promiseFunction = jest.fn(() => Promise.resolve("PROMISED"));

        TestRenderer.create(
          <Repromised promise={promiseFunction}>{children}</Repromised>
        );

        setTimeout(() => {
          expect(children.mock.calls[1][0].status).toBe(Status.Fulfilled);
          expect(children.mock.calls[1][0].value).toBe("PROMISED");
          expect(children.mock.calls[1][0].error).toBe(null);
          expect(children.mock.calls[1][0].isLoading).toBe(false);
          expect(children.mock.calls[1][0].requireValue).toBe("PROMISED");

          done();
        }, 5);
      });
    });

    describe("when props.promise rejects with an error", () => {
      test("calls props.children with the snapshot which expresses the promise is pending at 1st call", done => {
        const children = jest.fn(_ => null);
        const error = new Error("REJECTED");
        const promiseFunction = jest.fn(() => Promise.reject<any>(error));

        TestRenderer.create(
          <Repromised promise={promiseFunction}>{children}</Repromised>
        );

        setTimeout(() => {
          expect(children.mock.calls[0][0].status).toBe(Status.Pending);
          expect(children.mock.calls[0][0].value).toBe(null);
          expect(children.mock.calls[0][0].error).toBe(null);
          expect(children.mock.calls[0][0].isLoading).toBe(true);
          expect(() => children.mock.calls[0][0].requireValue).toThrow();

          done();
        }, 5);
      });

      test("calls props.children with the snapshot which expresses the promise is rejected at 2st call", done => {
        const children = jest.fn(_ => null);
        const error = new Error("REJECTED");
        const promiseFunction = jest.fn(() => Promise.reject<any>(error));

        TestRenderer.create(
          <Repromised promise={promiseFunction}>{children}</Repromised>
        );

        setTimeout(() => {
          expect(children.mock.calls[1][0].status).toBe(Status.Rejected);
          expect(children.mock.calls[1][0].value).toBe(null);
          expect(children.mock.calls[1][0].error).toBe(error);
          expect(children.mock.calls[1][0].isLoading).toBe(false);
          expect(() => children.mock.calls[1][0].requireValue).toThrow();

          done();
        }, 5);
      });
    });
  });
});
