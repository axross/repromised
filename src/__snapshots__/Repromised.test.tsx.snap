// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repromised> has props.children() as a render-props resolving a promise 1`] = `
<span>
  Symbol(returnValue)
</span>
`;

exports[`<Repromised> renders null if props.children is void 1`] = `null`;
